### PR TITLE
Cache auth scheme resolution

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

Cache deterministic auth scheme resolution results to reduce per-request CPU overhead. Auth scheme resolution is called on every request and produces the same result for a given set of inputs when using the default provider - caching avoids repeated allocation and computation.

## Modifications

- Generate `ConcurrentHashMap<String, List<AuthSchemeOption>> authSchemeCache` field on sync/async client classes
- Cache activates only when: 1) no per-request auth scheme provider override AND 2) client-level provider `instanceof DefaultXxxAuthSchemeProvider`
- Disabled for endpoint-based auth provider services (e.g. S3) where resolution depends on dynamic endpoint parameters
- Cache key composition varies by service characteristics:
  - Uniform-auth + sigv4: `region`
  - Uniform-auth + sigv4a: `region + ":" + regionSet`
  - Per-op-auth + sigv4: `operationName + ":" + region`
  - Per-op-auth + sigv4a: `operationName + ":" + region + ":" + regionSet`

## Testing

- Added unit tests
- Ran benchmarks confirming no regression on full-roundtrip workloads and significant savings on the isolated auth resolution path

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java [hunk 1]
```diff
@@ -393,13 +395,33 @@ static MethodSpec resolveAuthSchemeOptionsMethod(AuthSchemeSpecUtils authSchemeS
                              SdkInternalExecutionAttribute.class,
                              "Expected an instance of " + authSchemeSpecUtils.providerInterfaceName().simpleName());
 
+        // Use cache for simple (non-endpoint-based, non-S3) auth scheme resolution when using the default provider
+        boolean canCache = !authSchemeSpecUtils.useEndpointBasedAuthProvider() && !endpointRulesSpecUtils.isS3();
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: s3 should already have useEndpointBasedAuthProvider right?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java [hunk 2]
```diff
@@ -414,6 +436,24 @@ static MethodSpec resolveAuthSchemeOptionsMethod(AuthSchemeSpecUtils authSchemeS
...
 
+    /**
+     * Returns a field spec for the auth scheme options cache, used when simple (non-endpoint-based) auth is in effect.
+     */
+    static Optional<FieldSpec> authSchemeCacheField(AuthSchemeSpecUtils authSchemeSpecUtils,
+                                                    EndpointRulesSpecUtils endpointRulesSpecUtils) {
+        if (authSchemeSpecUtils.useEndpointBasedAuthProvider() || endpointRulesSpecUtils.isS3()) {
+            return Optional.empty();
+        }
+        ClassName concurrentHashMap = ClassName.get("java.util.concurrent", "ConcurrentHashMap");
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java:439]: A few thoughts on the currentHashMap - I believe a very small number of services have any per-operation auth defined (I found 12) and some services can have a large number of operations that all share the same auth (EC2 with... 800ish operations) which means we could potentially ending up storing a pretty large map in some cases which could impact memory usage, without any added value from it. 

What about either:
1) adding an additional constraint on caching and using caching only on services that do not define per-operation auth. or.
2) Special casing services with per-operation auth and using a map only for those.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java [hunk 3]
```diff
@@ -393,13 +395,33 @@ static MethodSpec resolveAuthSchemeOptionsMethod(AuthSchemeSpecUtils authSchemeS
...
         if (authSchemeSpecUtils.useEndpointBasedAuthProvider()) {
             addEndpointBasedAuthSchemeResolution(builder, authSchemeSpecUtils, endpointRulesSpecUtils);
         } else {
             addSimpleAuthSchemeResolution(builder, authSchemeSpecUtils);
         }
 
-        if (endpointRulesSpecUtils.isS3()) {
+        if (canCache) {
+            builder.beginControlFlow("if (useCache)");
+            builder.addStatement("authSchemeCache.put(operationName, options)");
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: I think this relies on an assumption that the resolver always returns an immutable list - which is true right now for our default auth resolver (that we already check for), but it seems like a possible place we could mess up long term.  I'm wondering if we chould either just wrap in Collections.unmodifiableList or do a check if it already is and if not, wrap as a defensive check....
Or if not, then we should add a test on the generated auth resolvers that enforces the immutablility and comments about why we depend on it?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java [hunk 4]
```diff
@@ -393,13 +395,33 @@ static MethodSpec resolveAuthSchemeOptionsMethod(AuthSchemeSpecUtils authSchemeS
                              SdkInternalExecutionAttribute.class,
                              "Expected an instance of " + authSchemeSpecUtils.providerInterfaceName().simpleName());
 
+        // Use cache for simple (non-endpoint-based, non-S3) auth scheme resolution when using the default provider
+        boolean canCache = !authSchemeSpecUtils.useEndpointBasedAuthProvider() && !endpointRulesSpecUtils.isS3();
+        if (canCache) {
+            ClassName defaultProviderClass = authSchemeSpecUtils.defaultAuthSchemeProviderName();
+            builder.addStatement("boolean useCache = requestAuthSchemeProvider == null "
+                                 + "&& authSchemeProvider instanceof $T", defaultProviderClass);
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: What happens if there is a request level override of, say region?  Would we use the correct overriden signing region?

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java (hunk 1 above)
Comment (davidh44) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: Yes, removed redundant isS3() check

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java (hunk 4 above)
Comment (davidh44) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: This isn't possible from normal request overrides via `AwsRequestOverrideConfiguration`. `AwsExecutionContextBuilder` sets `AWS_REGION` from only client config.

The edge case is a custom interceptor setting the `ExecutionAttribute`. To handle that, updated to include `region` and `regionSet` (conditionally) in the cache key

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/auth/AuthSchemeCacheTest.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.protocolrestjson.ProtocolRestJsonClient;
+import software.amazon.awssdk.services.protocolrestjson.auth.scheme.ProtocolRestJsonAuthSchemeProvider;
+
+/**
+ * Tests verifying auth scheme options cache behavior:
+ * - Custom client-level providers bypass the cache (instanceof guard)
+ * - Per-request overrides bypass the cache
+ * - Default provider with caching doesn't break correctness
+ */
```
Comment (S-Saranya1) [inline test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/auth/AuthSchemeCacheTest.java]: Worth adding a few more test cases covering:
   - regionSet varies per request
   - different operations get their own results
   - cache hit (default provider resolves only once for repeated same-operation calls)

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java (hunk 4 above)
Comment (S-Saranya1) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: Yeah, region and regionSet both have this issue, worth checking this.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java (hunk 3 above)
Comment (davidh44) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java]: Updated to always wrap with unmodifiableList

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/auth/AuthSchemeCacheTest.java (hunk 5 above)
Comment (davidh44) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/auth/AuthSchemeCacheTest.java]: Added test cases to cover these

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java (hunk 2 above)
Comment (davidh44) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/ClientClassUtils.java:439]: Good callout, updated cache key to be constructed by parts conditionally:  `operation` (hasPerOperationAuthSchemesOverrides), `region` (SigV4), `regionSet` (SigV4a)

**Review by alextwoods — APPROVED** [review]
